### ISSUE TITLE
GUI Mockup: History/Timeline

### ISSUE THREAD

**Issue by star-buck** [banji-project/ring-issues#14]
could look like this:
![history](https://cloud.githubusercontent.com/assets/840070/25556064/2258853a-2cf6-11e7-959d-f120b78e96de.png)

- sidebar shows last called contacts
- main area shows contact info and call/chat timeline

**Comment by Elv13** [comment]
Here's the first half

![image](https://cloud.githubusercontent.com/assets/340384/25899767/16124b82-355f-11e7-8aaa-5d8831503198.png)

![image](https://cloud.githubusercontent.com/assets/340384/25899784/248d8cf8-355f-11e7-9e19-bacd253d0bc7.png)

This is the "peers timeline" or "friendlist". It gets re-ordered and replaces the CallList in the timeline GUI. It performs mostly as expected, the recent contacts bubble to the top and so on.

While the other Ring clients had a seemingly similar mode, after inspecting the code, it turned out to be unsuitable for developing a timeline mode. I had to start from scratch building a true timeline that could scale to your requirements. The second half of the timeline UX is the right pane in your screenshot above. I am about 65% done writing the backend for it. This is actually more complicated than [it sounds](https://xkcd.com/1425/) as a contact can have multiple phone numbre / RingID, there can be conversations and other, external, media in the mix. Upstream Ring does not have this feature at all, so this is "uncharted" territory. It is/was planned and I wrote some of the necessary code years ago, but never actually got to the point where a true timeline was possible.

It is still not ready to be released. During the next week it should stabilize to the point where it become usable. Until major delays arrive in the next few days, I estimate it will be delivered on time as agreed. As most of the work in under the hood, the pieces should fall into place near the end of next week.

**Comment by Elv13** [comment]
(based on the emailed requirements and screenshot above) 

 * [X] Implement contact method deduplication when registered ring names are used
 * [X] Update the phonebook/phonedirectory to support registered ring names
 * [X] Write a new Peers Timeline Model
 * [X] Implement a GUI for the Peers Timeline Model
 * [X] Upgrade the ContactMethod class to support live attributes such as active calls, recording and active media (previously only available in the call scope)
 * [X] Refactor the lower level APIs to embed timeline metadata deeper into the object tree
 * [ ] Create a new Person TimeLine Model (in progress)
     * [X] Track text messages timeline
     * [X] Track recordings
     * [ ] Track contact edition / changes (?) (postponed)
     * [ ] Track screenshots
     * [X] Track calls and videos
 * [X] Complete the text message group feature
 * [ ] Merge the texts when the person has multiple "sources" like a different RingID for desktop and mobile or a real phone number + SMS (partially done)
 * [ ] Fix multi-party conversations (broken upstream) (postponed)
 * [X] Fix multi party video (SEGFAULT in the daemon...)
 * [ ] <s>Port the mobile GUI to the desktop version and add the remaining timeline features</s> (unsalvageable)
 * [ ] Add context menu
 * [X] Add the ability to delete individual elements
 * [X] Create a nicer widget to send messages with HTML (bold, underline, italic) and smiley support
 * [ ] Ability to search/filter the timeline (? I am not sure if this is required yet, up to you)
 * [X] Think ahead and plan the day when data will be synchronized between devices
 * [X] Create "pretty" chat bubbles to add visibility to the message direction...
 * [X] Detect new messages and remote status
 * [X] Write a player widget for the timeline audio media (recordings, voicemails)
 * [X] Make the ring-lrc media subsystem QML aware
 * [ ] Try harder to deduplicate the timeline
 * [ ] Fix all bugs, it is currently so buggy it is unusable

**Comment by Elv13** [comment]
[Weekly update]

Lot of progress, but still not winning the Skype "pretty UX crown" anytime soon. This is a lot of work, bug fixing/debugging and related tasks. I would have liked to have a completed prototype for today, but given how buggy it currently is, I think it wont be before Friday before it's ready for serious testing.

Like last week, please ignore the color, ugliness and debugging GUI. Most elements are now tracked correctly and have widgets. I still need to merge all that into the unified timeline and fix some duplication bugs. Anyway, some screenshots:

![image](https://cloud.githubusercontent.com/assets/340384/26191659/c5e20732-3b7c-11e7-8997-5fcf92541df5.png)

![image](https://cloud.githubusercontent.com/assets/340384/26191666/cbc49a70-3b7c-11e7-8cf0-d18ab35d9c11.png)

![image](https://cloud.githubusercontent.com/assets/340384/26191672/d1557f22-3b7c-11e7-8356-0fb80fbf752d.png)

![image](https://cloud.githubusercontent.com/assets/340384/26191676/d43f82aa-3b7c-11e7-8fe0-07ae7ca458a1.png)

Edit:
![pick13](https://cloud.githubusercontent.com/assets/340384/26238271/61fae930-3c46-11e7-9db2-50e6e5d233d7.png)


Questions:

 * Should I integrate the video dialog in the contact view or should I keep it as a separated window?
 * Do you want rich text messages (bold, italic, links, bullets, etc) or only text+emoji

**Comment by star-buck** [comment]
Really nice screenshots.
Lets keep the videocall window seperate for now.
Rich text would be fine, even with images or small video snippets if a user wants to record and send a video message to one of the contacts, but that all can be added later. 

[One idea we could even do in the mid-future is adding IMAP Email mapping autosorted in the same visual way like sms-text-chat, that is left and right, even with multiple people involved (like a groupchat in whatsapp). So "me" would always be on the right side and all other responses or emails on the left.]

**Comment by Elv13** [comment]
[1 day late...] Weekly update

I just finished the "main development". It took a bit more time than I expected, but the whole thing (and implementation) was rather ambitious to begin with. The look isn't finished yet, but at least most of the uglyness is gone.

First of all, the timeline itself is now implemented. The following media are supported (to various degrees, including "not yet, but there is some code for it"):

* TEXT_MESSAGE_GROUP (by session, time category or timeout) 
* TEXT_MESSAGE (with Emoji, rich text isn't enabled in the other clients yet, but is supported)
* TIME_CATEGORY  (today, yesterday, last week, ...)
* CALL_GROUP  (consecutive calls to the same person)
* CALL             
* AUDIO_RECORDING
* SCREENSHOT (work in progress)
* EMAIL (planned)

Everything is indexed (I hope) correctly in an unified timeline:

![image](https://cloud.githubusercontent.com/assets/340384/26518433/73901bd2-427e-11e7-85d0-1cbf7e94d267.png)

![image](https://cloud.githubusercontent.com/assets/340384/26518600/5c2f1698-4281-11e7-85f7-f5a366743ee1.png)

Also mostly implemented is the timeline navigation feature when you overlay either the people timeline or the media timeline scrollbar:

![image](https://cloud.githubusercontent.com/assets/340384/26518439/ab172d2a-427e-11e7-973f-bb7b13403b4c.png)

There is still a fair amount of bugs, however none should take too much time to fix. Initial testing is now welcome (@shadeslayer please create a new build on Monday, thanks). As I said, there is still some known bugs. But report are welcome, but I am already working on fixing the obvious ones.

Feel free to bash the design, everything can be changed, it is fairly modular and most of the complexity is in the backend, not the UX. There will be some extra polishing if you are satisfied with the general look. I left it a bit rough to make sure you like it before spending time on pixel-perfectness. 

> [One idea we could even do in the mid-future is adding IMAP Email mapping autosorted in the same visual way like sms-text-chat, that is left and right, even with multiple people involved (like a groupchat in whatsapp). So "me" would always be on the right side and all other responses or emails on the left.]

Yes, that would be very cool. Turning the project into some sort of "BlackBerry Hub app" where you can unify the different media sources. Ring-LRC is designed for that, it already supports 16 sources, more can be added without touching the "core code". Of course it all depends on having a groupware solution (like Akonadi[-ng], kimap, *whatever works* or Evolution data Server). Writing a full IMAP implementation would be madness (well, if it's what you want, I can do it anyway, I just doubt there is a good return on investment ;) ).

**Comment by star-buck** [comment]
Looks really good, cant wait to test it. :)

**Comment by shadeslayer** [comment]
Updated the ring-kde and ring packages on DCI from the next branch. Just waiting on ARM64 to finish building.

**Comment by Elv13** [comment]
@shadeslayer thanks

Bugs to fix:

 * [ ] Complete the split of the call and IM UXes
    * [X] Only show the right windows
    * [ ] Turn the phone window into a dialog type
    * [ ] Add a systray entry for each mode
    * [ ] Add extra shortcuts and KCFG for each mode
    * [ ] Upgrade the configuration dialog to reflect that duality
 * [ ] Display the correct timeline when a new message arrives
 * [X] Highlight the PeersTimeline entry with awaiting new messages from this session
 * [X] Mitigate the QML double free when using QSharedPointer (https://bugreports.qt.io/browse/QTBUG-43080)
 * [ ] Fix the scollbar
 * [ ] Reload the categories every half hours (**IN PROGRESS**)
 * [X] Auto scroll down the chat
 * [X] Highlight the unread messages
 * [ ] Enable the audio recordings in the timeline by default
 * [ ] Only show one entry per person in the timeline (instead of per account)  (**IN PROGRESS**)
 * [ ] Squash the stats of all person account into 1 summary
 * [ ] Show more call metadata when clicking on them in the timeline
 * [ ] If there is audio or video media in progress, show the media toolbar in the chat.
 * [X] Add more fields to the contact editor 
    * [X] Allow to change the contact photo in the GUI (rather than external apps)
    * [X] Add multi-account support in the contact details
    * [X] Add addresses to the contact editor
 * [ ] Enable the timeline scrollbar in the main timeline.
 * [X] Fix the segfault late when quitting the application
 * [ ] Fix all new memory leaks
 * [ ] Fix Clazy warnings
 * [ ] Retire the older code paths like the QtWidget CategorizedItemView that already have been rewritten in QML but still have unported instances.
 * [ ] Write more unit tests

*new bug report welcome*
